File: reference/fud-ai/ios/calorietracker/Services/MealShare.swift
import Foundation
import UIKit

/// Encodes/decodes a logged meal into a `fudai://add-meal?d=<base64url>` deep link so it
/// can be shared (AirDrop, Messages, any app) and imported directly into Fud AI — including
/// cross-platform. The payload schema is byte-identical to the Android `MealShare`, so a
/// link produced on one platform imports on the other.
enum MealShare {
    static let scheme = "fudai"
    static let host = "add-meal"
    /// Universal Link host + path — tapping this in a messenger opens the app directly
    /// (verified via /.well-known/apple-app-site-association), no browser.
    static let webHost = "www.fud-ai.app"
    static let webPath = "/add-meal"
    private static let version = 1

    // MARK: - Encode

    /// A shareable link carrying every entry's nutrients (no image). Uses an https://fud-ai.app
    /// URL so messengers (WhatsApp etc.) make it tappable — the page then opens the app via the
    /// fudai://add-meal scheme. `d` is byte-identical to the scheme link, so import is unchanged.
    static func link(for entries: [FoodEntry]) -> URL? {
        let payload: [String: Any] = ["v": version, "meals": entries.map(mealDict)]
        guard JSONSerialization.isValidJSONObject(payload),
              let data = try? JSONSerialization.data(withJSONObject: payload),
              !data.isEmpty else { return nil }
        var comps = URLComponents()
        comps.scheme = "https"
        comps.host = webHost
        comps.path = webPath
        comps.queryItems = [URLQueryItem(name: "d", value: base64url(data))]
        return comps.url
    }

    /// Human-readable summary plus the import link — the text put on the share sheet.
    static func shareText(for entries: [FoodEntry]) -> String {
        var lines = entries.map { e -> String in
            let macros = "\(Int(e.protein.rounded()))P · \(Int(e.carbs.rounded()))C · \(Int(e.fat.rounded()))F"
            let prefix = e.emoji.map { "\($0) " } ?? ""
            return "\(prefix)\(e.name) — \(e.calories) kcal · \(macros)"
        }
        if let link = link(for: entries) {
            lines.append("")
            lines.append("Open in Fud AI to add:")
            lines.append(link.absoluteString)
        }
        return lines.joined(separator: "\n")
    }

    private nonisolated static func mealDict(_ e: FoodEntry) -> [String: Any] {
        var d: [String: Any] = [
            "name": e.name,
            "calories": e.calories,
            "protein": e.protein,
            "carbs": e.carbs,
            "fat": e.fat,
            "mealType": e.mealType.rawValue,
        ]
        if let emoji = e.emoji { d["emoji"] = emoji }
        func put(_ key: String, _ v: Double?) { if let v { d[key] = v } }
        put("sugar", e.sugar); put("addedSugar", e.addedSugar); put("fiber", e.fiber)
        put("saturatedFat", e.saturatedFat); put("monounsaturatedFat", e.monounsaturatedFat)
        put("polyunsaturatedFat", e.polyunsaturatedFat); put("cholesterol", e.cholesterol)
        put("sodium", e.sodium); put("potassium", e.potassium); put("transFat", e.transFat)
        put("calcium", e.calcium); put("iron", e.iron); put("magnesium", e.magnesium); put("zinc", e.zinc)
        put("vitaminA", e.vitaminA); put("vitaminC", e.vitaminC); put("vitaminD", e.vitaminD)
        put("vitaminB12", e.vitaminB12); put("vitaminE", e.vitaminE); put("vitaminK", e.vitaminK)
        put("folate", e.folate); put("omega3", e.omega3)
        put("servingSizeGrams", e.servingSizeGrams)
        if let unit = e.selectedServingUnit { d["selectedServingUnit"] = unit }
        put("selectedServingQuantity", e.selectedServingQuantity)
        if let note = e.customNote { d["customNote"] = note }
        return d
    }

    // MARK: - Decode

    /// True for both the custom scheme and the https Universal Link that carry a shared meal.
    static func handles(_ url: URL) -> Bool {
        if url.scheme == scheme, url.host == host { return true }
        if url.scheme == "https",
           url.host == webHost || url.host == "fud-ai.app",
           url.path == webPath { return true }
        return false
    }

    /// Parse a shared-meal link (custom scheme or https Universal Link) back into fresh
    /// `FoodEntry` values (new ids, logged now).
    static func meals(from url: URL) -> [FoodEntry]? {
        guard handles(url),
              let comps = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let encoded = comps.queryItems?.first(where: { $0.name == "d" })?.value,
              let data = data(fromBase64url: encoded),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let mealsArr = obj["meals"] as? [[String: Any]]
        else { return nil }
        let entries = mealsArr.compactMap(entry(from:))
        return entries.isEmpty ? nil : entries
    }

    private nonisolated static func entry(from d: [String: Any]) -> FoodEntry? {
        guard let name = (d["name"] as? String), !name.isEmpty,
              let calories = (d["calories"] as? NSNumber)?.intValue else { return nil }
        func dbl(_ k: String) -> Double? { (d[k] as? NSNumber)?.doubleValue }
        let meal = MealType(rawValue: (d["mealType"] as? String) ?? "") ?? .currentMeal
        return FoodEntry(
            name: name,
            calories: calories,
            protein: dbl("protein") ?? 0,
            carbs: dbl("carbs") ?? 0,
            fat: dbl("fat") ?? 0,
            emoji: d["emoji"] as? String,
            source: .manual,
            mealType: meal,
            sugar: dbl("sugar"), addedSugar: dbl("addedSugar"), fiber: dbl("fiber"),
            saturatedFat: dbl("saturatedFat"), monounsaturatedFat: dbl("monounsaturatedFat"),
            polyunsaturatedFat: dbl("polyunsaturatedFat"), cholesterol: dbl("cholesterol"),
            sodium: dbl("sodium"), potassium: dbl("potassium"), transFat: dbl("transFat"),
            calcium: dbl("calcium"), iron: dbl("iron"), magnesium: dbl("magnesium"), zinc: dbl("zinc"),
            vitaminA: dbl("vitaminA"), vitaminC: dbl("vitaminC"), vitaminD: dbl("vitaminD"),
            vitaminB12: dbl("vitaminB12"), vitaminE: dbl("vitaminE"), vitaminK: dbl("vitaminK"),
            folate: dbl("folate"), omega3: dbl("omega3"),
            servingSizeGrams: dbl("servingSizeGrams"),
            servingUnitOptions: [],
            selectedServingUnit: d["selectedServingUnit"] as? String,
            selectedServingQuantity: dbl("selectedServingQuantity"),
            customNote: d["customNote"] as? String
        )
    }

    // MARK: - Share sheet

    /// Present the system share sheet directly via UIKit (avoids the blank-page issue of a
    /// SwiftUI `.sheet` nested inside another sheet).
    static func presentShareSheet(for entries: [FoodEntry]) {
        guard !entries.isEmpty else { return }
        let av = UIActivityViewController(activityItems: [shareText(for: entries)], applicationActivities: nil)
        guard let scene = UIApplication.shared.connectedScenes.first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene,
              let root = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController else { return }
        var top = root
        while let presented = top.presentedViewController { top = presented }
        av.popoverPresentationController?.sourceView = top.view
        av.popoverPresentationController?.sourceRect = CGRect(x: top.view.bounds.midX, y: top.view.bounds.midY, width: 0, height: 0)
        av.popoverPresentationController?.permittedArrowDirections = []
        top.present(av, animated: true)
    }

    // MARK: - base64url helpers

    private static func base64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func data(fromBase64url s: String) -> Data? {
        var b64 = s.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let pad = b64.count % 4
        if pad > 0 { b64 += String(repeating: "=", count: 4 - pad) }
        return Data(base64Encoded: b64)
    }
}

File: fud-ai/android/app/src/main/java/com/apoorvdarshan/calorietracker/services/MealShare.kt
package com.apoorvdarshan.calorietracker.services

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.apoorvdarshan.calorietracker.models.FoodEntry
import com.apoorvdarshan.calorietracker.models.FoodSource
import com.apoorvdarshan.calorietracker.models.MealType
import org.json.JSONArray
import org.json.JSONObject
import java.util.Base64
import kotlin.math.roundToInt

/**
 * Encodes/decodes a logged meal into a `fudai://add-meal?d=<base64url>` deep link so it can be
 * shared (any app) and imported directly into Fud AI — including cross-platform. The payload
 * schema is byte-identical to the iOS `MealShare`, so a link produced on one platform imports
 * on the other.
 */
object MealShare {
    const val SCHEME = "fudai"
    const val HOST = "add-meal"
    /** Verified App Link host + path — tapping this opens the app directly (no browser). */
    const val WEB_HOST = "www.fud-ai.app"
    const val WEB_PATH = "/add-meal"
    private const val VERSION = 1

    // MARK: Encode

    /**
     * A shareable link carrying every entry's nutrients (no image). Uses an https://fud-ai.app
     * URL so messengers (WhatsApp etc.) make it tappable — the page then opens the app via the
     * fudai://add-meal scheme. `d` is byte-identical to the scheme link, so import is unchanged.
     */
    fun link(entries: List<FoodEntry>): String {
        val meals = JSONArray()
        entries.forEach { meals.put(mealJson(it)) }
        val payload = JSONObject().put("v", VERSION).put("meals", meals)
        val b64 = Base64.getUrlEncoder().withoutPadding()
            .encodeToString(payload.toString().toByteArray(Charsets.UTF_8))
        return "https://$WEB_HOST$WEB_PATH?d=$b64"
    }

    /** Fire the system share sheet with a readable summary + the fudai://add-meal link. */
    fun share(context: Context, entries: List<FoodEntry>) {
        if (entries.isEmpty()) return
        val send = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_TEXT, shareText(entries))
        }
        context.startActivity(Intent.createChooser(send, "Share meal"))
    }

    /** Human-readable summary plus the import link — the text put on the share sheet. */
    fun shareText(entries: List<FoodEntry>): String {
        val lines = entries.map { e ->
            val macros = "${e.protein.roundToInt()}P · ${e.carbs.roundToInt()}C · ${e.fat.roundToInt()}F"
            val prefix = e.emoji?.let { "$it " } ?: ""
            "$prefix${e.name} — ${e.calories} kcal · $macros"
        }.toMutableList()
        lines.add("")
        lines.add("Open in Fud AI to add:")
        lines.add(link(entries))
        return lines.joinToString("\n")
    }

    private fun mealJson(e: FoodEntry): JSONObject {
        val d = JSONObject()
            .put("name", e.name)
            .put("calories", e.calories)
            .put("protein", e.protein)
            .put("carbs", e.carbs)
            .put("fat", e.fat)
            .put("mealType", e.mealType.name.lowercase())
        e.emoji?.let { d.put("emoji", it) }
        fun put(key: String, v: Double?) { if (v != null) d.put(key, v) }
        put("sugar", e.sugar); put("addedSugar", e.addedSugar); put("fiber", e.fiber)
        put("saturatedFat", e.saturatedFat); put("monounsaturatedFat", e.monounsaturatedFat)
        put("polyunsaturatedFat", e.polyunsaturatedFat); put("cholesterol", e.cholesterol)
        put("sodium", e.sodium); put("potassium", e.potassium); put("transFat", e.transFat)
        put("calcium", e.calcium); put("iron", e.iron); put("magnesium", e.magnesium); put("zinc", e.zinc)
        put("vitaminA", e.vitaminA); put("vitaminC", e.vitaminC); put("vitaminD", e.vitaminD)
        put("vitaminB12", e.vitaminB12); put("vitaminE", e.vitaminE); put("vitaminK", e.vitaminK)
        put("folate", e.folate); put("omega3", e.omega3)
        put("servingSizeGrams", e.servingSizeGrams)
        e.selectedServingUnit?.let { d.put("selectedServingUnit", it) }
        put("selectedServingQuantity", e.selectedServingQuantity)
        e.customNote?.let { d.put("customNote", it) }
        return d
    }

    // MARK: Decode

    /** True for both the custom scheme and the https App Link that carry a shared meal. */
    fun handles(uri: Uri): Boolean {
        if (uri.scheme == SCHEME && uri.host == HOST) return true
        return uri.scheme == "https" &&
            (uri.host == WEB_HOST || uri.host == "fud-ai.app") &&
            uri.path == WEB_PATH
    }

    /**
     * Parse a shared-meal link (custom scheme or https App Link) back into fresh [FoodEntry]
     * values (new ids, logged now).
     */
    fun meals(uri: Uri): List<FoodEntry>? {
        if (!handles(uri)) return null
        val encoded = uri.getQueryParameter("d") ?: return null
        val json = runCatching {
            val bytes = Base64.getUrlDecoder().decode(encoded)
            JSONObject(String(bytes, Charsets.UTF_8))
        }.getOrNull() ?: return null
        val mealsArr = json.optJSONArray("meals") ?: return null
        val entries = (0 until mealsArr.length()).mapNotNull { i ->
            mealsArr.optJSONObject(i)?.let(::entryFrom)
        }
        return entries.ifEmpty { null }
    }

    private fun entryFrom(d: JSONObject): FoodEntry? {
        val name = d.optString("name").takeIf { it.isNotEmpty() } ?: return null
        if (!d.has("calories")) return null
        fun dbl(k: String): Double? = if (d.has(k) && !d.isNull(k)) d.optDouble(k) else null
        val meal = runCatching { MealType.valueOf(d.optString("mealType").uppercase()) }
            .getOrDefault(MealType.currentMeal)
        return FoodEntry(
            name = name,
            calories = d.optInt("calories"),
            protein = d.optDouble("protein", 0.0),
            carbs = d.optDouble("carbs", 0.0),
            fat = d.optDouble("fat", 0.0),
            emoji = if (d.has("emoji")) d.optString("emoji") else null,
            source = FoodSource.MANUAL,
            mealType = meal,
            sugar = dbl("sugar"), addedSugar = dbl("addedSugar"), fiber = dbl("fiber"),
            saturatedFat = dbl("saturatedFat"), monounsaturatedFat = dbl("monounsaturatedFat"),
            polyunsaturatedFat = dbl("polyunsaturatedFat"), cholesterol = dbl("cholesterol"),
            sodium = dbl("sodium"), potassium = dbl("potassium"), transFat = dbl("transFat"),
            calcium = dbl("calcium"), iron = dbl("iron"), magnesium = dbl("magnesium"), zinc = dbl("zinc"),
            vitaminA = dbl("vitaminA"), vitaminC = dbl("vitaminC"), vitaminD = dbl("vitaminD"),
            vitaminB12 = dbl("vitaminB12"), vitaminE = dbl("vitaminE"), vitaminK = dbl("vitaminK"),
            folate = dbl("folate"), omega3 = dbl("omega3"),
            servingSizeGrams = dbl("servingSizeGrams"),
            selectedServingUnit = if (d.has("selectedServingUnit")) d.optString("selectedServingUnit") else null,
            selectedServingQuantity = dbl("selectedServingQuantity"),
            customNote = if (d.has("customNote")) d.optString("customNote") else null
        )
    }
}
